README
netcalc  Copyright (c) 2021  Matthew Reath

Overview
========

netcalc is an advanced CLI-based network subnetting tool. It provides the
following features:

	- provided an ip address and mask, it will calculate the network
	  address, broadcast address, bits in mask, and all hosts 
	  of the network (done)

	- provided an ip address, mask, and network count, it will 
	  calculate the network addresses, and masks in order to
	  subnet the network into smaller networks based on the
	  number of networks desired (done)

	- provided an ip address, mask, and host count, it will
	  calculate the network addresses, and mask in ord to
	  subnet the network into smaller networks based on the
	  number of hosts desired (done)

	- provided an ip address, mask, and a list of host count
	  values, it will calculate a list of VLSM subnets that
	  match the list of host count requirements. Instead of 
	  making the subnets all the same size, it only creates
	  enough subnets to match the list, and they may all be
	  different sizes (done)

	- provided a list of subnets, it will try to summarize
	  the subnets into one (or more) supernets if possible. (done)

netcalc uses bit shifting and mask operators to work with ip addresses
in their actual 32 bit integer form. It converts from dotted decimal
notation to unsigned integer values, and then manipulates the integer
values.

netcalc also uses a binary tree structure to subnet networks by 
extending the mask by one bit and create two new networks as a nodes
left and right child. Left is the lower half, right the upper half.

Compiling/Installing
====================

autoreconf
automake --add-missing
./configure
make
make install

Usage
=====

netcalc --help
Usage:
  netcalc [OPTION?] - calculate network information

Help Options:
  -?, --help               Show help options

Application Options:
  -v, --verbose            Use verbose output
  -h, --hosts=HOSTS        Specify the host count to use to subnet the network [IPv4 only]
  -n, --nets=NETS          Specify the net count to use to subnet the network
  -l, --vlsm=VLSM_LIST     Comma seperated list of host count requirements for VLSM network
  -s, --summary            Summarize a list of subnets into a one or more supernets
  -i, --ip-address         Specific the IPv4 Address to use for calculation
  -m, --netmask            Specific the IPv4 Network Mask to use for calculation


Basic network information
=========================

netcalc -v -i 10.1.1.0 -m 255.255.255.252
netcalc 0.8  Copyright (C) 2017 Matt Reath
This program comes with ABSOLUTELY NO WARRANTY;
This is free software, and you are welcome to redistribute it
under certain circumstances. See the included COPYING file
for more information.

Network:  10.1.1.0
Broadcast:  10.1.1.3
Mask:   255.255.255.252[/30]
Hosts:    2
10.1.1.1  255.255.255.252
10.1.1.2  255.255.255.252

Subnet a network based on hosts needed per subnet
=================================================
50 hosts per subnet needed, starting with 10.1.1.0/24

netcalc -v -h 50 -i 10.1.1.0 -m 255.255.255.0
netcalc 0.8  Copyright (C) 2017 Matt Reath
This program comes with ABSOLUTELY NO WARRANTY;
This is free software, and you are welcome to redistribute it
under certain circumstances. See the included COPYING file
for more information.

* = assigned network
[n] = # of useable hosts
+ = useable network

__10.1.1.0/24
 |__10.1.1.0/25
 | |__10.1.1.0/26*[62]+
 | |__10.1.1.64/26*[62]+
 |__10.1.1.128/25
 | |__10.1.1.128/26*[62]+
 | |__10.1.1.192/26*[62]+
 
 Subnet a network based on the number of networks needed
 =======================================================
 8 networks needed, starting with 10.1.1.0/24
 
  netcalc -v -n 8 -i 10.1.1.0 -m 255.255.255.0
  netcalc 0.8  Copyright (C) 2017 Matt Reath
  This program comes with ABSOLUTELY NO WARRANTY;
  This is free software, and you are welcome to redistribute it
  under certain circumstances. See the included COPYING file
  for more information.
  
  * = assigned network
  [n] = # of useable hosts
  + = useable network
  
  __10.1.1.0/24
   |__10.1.1.0/25
   | |__10.1.1.0/26
   | | |__10.1.1.0/27*[30]+
   | | |__10.1.1.32/27*[30]+
   | |__10.1.1.64/26
   | | |__10.1.1.64/27*[30]+
   | | |__10.1.1.96/27*[30]+
   |__10.1.1.128/25
   | |__10.1.1.128/26
   | | |__10.1.1.128/27*[30]+
   | | |__10.1.1.160/27*[30]+
   | |__10.1.1.192/26
   | | |__10.1.1.192/27*[30]+
   | | |__10.1.1.224/27*[30]+
  
  Subnet a network using VLSM
  ===========================
  2 - 2 host networks needed (point-to-point links)
  2 - 50 host networks needed
  Start with 10.1.1.0/24
  
  netcalc -v -l 2,2,50,50 -i 10.1.1.0 -m 255.255.255.0
  netcalc 0.8  Copyright (C) 2017 Matt Reath
  This program comes with ABSOLUTELY NO WARRANTY;
  This is free software, and you are welcome to redistribute it
  under certain circumstances. See the included COPYING file
  for more information.
  
  * = assigned network
  [n] = # of useable hosts
  + = useable network
  
  __10.1.1.0/24
   |__10.1.1.0/25
   | |__10.1.1.0/26
   | | |__10.1.1.0/27
   | | | |__10.1.1.0/28
   | | | | |__10.1.1.0/29
   | | | | | |__10.1.1.0/30*[2]+
   | | | | | |__10.1.1.4/30*[2]+
   | | | | |__10.1.1.8/29+
   | | | |__10.1.1.16/28+
   | | |__10.1.1.32/27+
   | |__10.1.1.64/26*[62]+
   |__10.1.1.128/25
   | |__10.1.1.128/26*[62]+
   | |__10.1.1.192/26+
  
  Supernet/summarize a set of networks
  ====================================

  more test.txt
  192.168.0.0 255.255.255.128
  192.168.0.128 255.255.255.128
  192.168.1.0 255.255.255.0
  192.168.2.0 255.255.255.0
  192.168.3.0 255.255.255.0

  
  netcalc -vs < test.txt
  Networks were summarized as follows:

  * = assigned network
  [n] = # of useable hosts
  + = useable network

  __192.168.0.0/22
   |__192.168.0.0/23
   | |__192.168.0.0/24
   | | |__192.168.0.0/25+
   | | |__192.168.0.128/25+
   | |__192.168.1.0/24+
   |__192.168.2.0/23
   | |__192.168.2.0/24+
   | |__192.168.3.0/24+
  


ChangeLog
commit 8b160f7f07abd7c6b30f597a16e477b2e2221aeb
Merge: b6513b6 0cc7ae6
Author: Matt Reath <mattreath@icloud.com>
Date:   Thu Apr 29 15:38:14 2021 -0500

    Merge pull request #13 from mpreath/invalid-mask-handling
    
    Invalid mask handling

commit 0cc7ae699f7f65ff861733e2439a89b4767caf06
Author: Matt Reath <mattreath@icloud.com>
Date:   Thu Apr 29 15:37:46 2021 -0500

    updated changelog

commit 1ab57bff012392cc5560f2f177c10c88ecff3b65
Author: Matt Reath <mattreath@icloud.com>
Date:   Thu Apr 29 15:34:37 2021 -0500

    set smallest mask possible at /8 or 255.0.0.0

commit b6513b6418104cf3e09ae3f95039bf0b0baad1f7
Merge: 8a337d6 a5fa523
Author: Matt Reath <mattreath@icloud.com>
Date:   Thu Apr 29 14:06:00 2021 -0500

    Merge pull request #12 from mpreath/optimize-summarization
    
    Optimize summarization

commit a5fa52365554612154cec16f58f8a3dc23e0a17f
Author: Matt Reath <mattreath@icloud.com>
Date:   Thu Apr 29 14:05:31 2021 -0500

    added a verbose option

commit 0012e00d7131cff671720a682294a305ce70978e
Author: Matt Reath <mattreath@icloud.com>
Date:   Thu Apr 29 14:00:39 2021 -0500

    remplaced array of structs with array of guint32

commit 7fa81b387afd510719e59521ab9ede263375ba2d
Author: Matt Reath <mattreath@icloud.com>
Date:   Thu Apr 29 13:43:51 2021 -0500

    Fixed a potential app breaking bug

commit c51344957d7464fee48923704ea4fb13c063d554
Author: Matt Reath <mattreath@icloud.com>
Date:   Thu Apr 29 12:47:05 2021 -0500

    implemented new O(n) summarization algorithm

commit e4ac28709496fe702d6afcaa61be8a421b1fd08a
Author: Matt Reath <mattreath@icloud.com>
Date:   Thu Apr 29 08:23:25 2021 -0500

    started new summarization algorithm

commit 3c7e2805b3971a1700ee00929145353ee9f30c2b
Author: Matt Reath <mattreath@icloud.com>
Date:   Thu Apr 29 00:41:47 2021 -0500

    pseudocode updated algorithm

commit 0c9a1255fc4dccf36e856d21bb1a1c424dff2910
Author: Matt Reath <mattreath@icloud.com>
Date:   Wed Apr 28 22:44:43 2021 -0500

    summarization tree now utilizes the heap - malloc

commit 8a337d6a4ead3a7426ff9682cd0db2f3393773f1
Merge: 7695baf 9f2bf22
Author: Matt Reath <mattreath@icloud.com>
Date:   Tue Apr 27 21:35:35 2021 -0500

    Merge pull request #11 from mpreath/error-handling-and-refinement
    
    Error handling and refinement

commit 9f2bf227fa7c1231da8c5dfb4197831ae53b4196
Merge: b98fd90 7695baf
Author: Matt Reath <mattreath@icloud.com>
Date:   Tue Apr 27 21:35:28 2021 -0500

    Merge branch 'master' into error-handling-and-refinement

commit b98fd90acf921d0c6d4c3a632974ec7f4b942635
Author: Matt Reath <mattreath@icloud.com>
Date:   Tue Apr 27 21:34:19 2021 -0500

    cleaned up unused code

commit 5243fa7e889da3ecfa0bc030bd25ba2993c89760
Author: Matt Reath <mattreath@icloud.com>
Date:   Tue Apr 27 19:04:07 2021 -0500

    cleaned up print_network_info

commit 30f0f96d9b392be08621c178a6848e64dbd9e7d3
Author: Matt Reath <mattreath@icloud.com>
Date:   Tue Apr 27 08:52:19 2021 -0500

    memory management fixes

commit 2cdbd32ba3fdfe790771fa73730af5f99fa0f0aa
Author: Matt Reath <mattreath@icloud.com>
Date:   Tue Apr 27 08:41:57 2021 -0500

    cleaned up help output

commit c0bc229dd53d45c32318d4fe97c532a2d59ff043
Author: Matt Reath <mattreath@icloud.com>
Date:   Tue Apr 27 08:36:48 2021 -0500

    added version option, cleaned up print_info

commit 0f9267dc0564651855c2c610d32fc9c3d50f4a42
Author: Matt Reath <mattreath@icloud.com>
Date:   Tue Apr 27 07:18:23 2021 -0500

    updated version number

commit b4b7009a110a27fc2f5e1556243992ce0471d843
Author: Matt Reath <mattreath@icloud.com>
Date:   Tue Apr 27 07:16:43 2021 -0500

    updated copyright statements; added notice to host.c

commit 66dbbbc9ccff20bd5d8be56a33fa9c7b37a73493
Author: Matt Reath <mattreath@icloud.com>
Date:   Tue Apr 27 07:08:15 2021 -0500

    rremoved an unused variable

commit 7695bafd5072e359cc67514687d25a32d02c5dec
Author: Matt Reath <mattreath@icloud.com>
Date:   Mon Apr 26 08:09:07 2021 -0500

    removed unused variable

commit 7aa5f4f029d63b9d6d6b81b7248571993c31efa8
Merge: efa0b55 a04a7aa
Author: Matt Reath <mattreath@icloud.com>
Date:   Mon Apr 26 07:16:57 2021 -0500

    Merge pull request #10 from mpreath/cidr-prefix-support
    
    Cidr prefix support

commit a04a7aa940956af9481ed0df3dac401edcee4049
Author: Matt Reath <mattreath@icloud.com>
Date:   Mon Apr 26 07:16:19 2021 -0500

    moved print_info() to not display when program fails

commit b30c9b3c03f5f84f361ecb23aa64821cebe86a10
Author: Matt Reath <mattreath@icloud.com>
Date:   Mon Apr 26 07:08:15 2021 -0500

    added handling of cidr to ip_address command parameter

commit 46f883be64cfc992dfdc763c9b052c79cd86c7a1
Author: Matt Reath <mattreath@icloud.com>
Date:   Sun Apr 25 22:57:30 2021 -0500

    added verbose check in network summarization

commit f141b78c61e232d68ea851b47370fd7fef954ccd
Author: Matt Reath <mattreath@icloud.com>
Date:   Sun Apr 25 22:50:19 2021 -0500

    cidrtoint(), initialize hosts and network for summarization

commit 0c22fba69aaa70595f44a2b9b3b7427fda9f7b82
Author: Matt Reath <mattreath@icloud.com>
Date:   Sun Apr 25 21:14:13 2021 -0500

    getline(), check for cidr, and cidr split

commit efa0b550d3ca326a87bb995f3ffa7810b5e34d58
Merge: 13bb367 0863cb1
Author: Matt Reath <mattreath@icloud.com>
Date:   Sun Apr 18 20:22:11 2021 -0500

    Merge pull request #9 from mpreath/update-version-copyright
    
    Updated version and copyright

commit 0863cb104810eb2e2682683a0002f56cad1243af
Author: Matt Reath <mattreath@icloud.com>
Date:   Sun Apr 18 20:21:11 2021 -0500

    Updated version and copyright

commit 13bb3679d58e002707de91df5d2aa00f7b029eeb
Merge: 8351a30 27d5abd
Author: Matt Reath <mattreath@icloud.com>
Date:   Sun Apr 18 20:15:47 2021 -0500

    Merge pull request #8 from mpreath/reformat-all-files
    
    Reformat all source files

commit 27d5abd27e228ae115b81f6af9c9b833e4423889
Author: Matt Reath <mattreath@icloud.com>
Date:   Sun Apr 18 20:15:16 2021 -0500


src/netcalc.c
/* 
    netcalc - a simple CLI subnet calculator written in ANSI C
    netcalc.c - main entry point into the netcalc application

    Copyright (c) 2021  Matthew Reath 

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
#include <config.h>
#include <utility.h>
#include <stdio.h>
#include <stdlib.h>
#include <host.h>
#include <network.h>
#include <network_tree.h>
#include <string.h>
#include <glib.h>
#include <glib/gprintf.h>

#define MAX_NUM_TREES 2000000

void print_info();
void print_usage();
void net_info(char *ip_address, char *mask);
void host_tree(char *ip_address, char *mask, int hosts);
void net_tree(char *ip_address, char *mask, int nets);
void vlsm_tree(char *ip_address, char *mask, char *nets);
void net_summary();

static char *vlsm_counts = NULL;
static char *ip_address = NULL;
static char *netmask = NULL;
static gint do_host_count = 0;
static gint do_net_count = 0;
static gboolean do_summary = FALSE;
static gboolean verbose = FALSE;
static gboolean version = FALSE;
//static gboolean do_help = FALSE;
static GOptionEntry entries[] =
	{
		{"verbose", 'v', 0, G_OPTION_ARG_NONE, &verbose,
		 "Use verbose output", NULL},
		 {"version", 'V', 0, G_OPTION_ARG_NONE, &version,
		 "Display version information", NULL},
		{"hosts", 'o', 0, G_OPTION_ARG_INT, &do_host_count,
		 "Specify the host count to use to subnet the network",
		 "HOSTS"},
		{"nets", 'n', 0, G_OPTION_ARG_INT, &do_net_count,
		 "Specify the net count to use to subnet the network",
		 "NETS"},
		{"vlsm", 'l', 0, G_OPTION_ARG_STRING, &vlsm_counts,
		 "Comma seperated list of host counts for VLSM network",
		 "VLSM_LIST"},
		{"summary", 's', 0, G_OPTION_ARG_NONE, &do_summary,
		 "Summarize subnets into one or more supernets",
		 NULL},
		{NULL}};

int main(int argc, char *argv[])
{

	GError *error = NULL;
	GOptionContext *context;
	context = g_option_context_new("[<ip_address/cidr> | <ip_address> <mask>] - calculate network information");
	g_option_context_add_main_entries(context, entries, NULL);
	if (!g_option_context_parse(context, &argc, &argv, &error))
	{
		g_print("option parsing failed: %s\n", error->message);
		exit(1);
	}

	if (argc == 3)
	{
		ip_address = (gchar *) malloc(1 + sizeof(gchar) * strlen(argv[argc - 2]));
		netmask = (gchar *) malloc (1 + sizeof(gchar) * strlen(argv[argc - 1])); 
		g_stpcpy(ip_address, argv[argc - 2]);
		g_stpcpy(netmask, argv[argc - 1]);

	} else if (argc == 2)
	{
		gchar** split_values = g_strsplit(argv[argc - 1], "/", 2);
		if(split_values[0] != NULL)
		{
			ip_address = (gchar *) malloc(1 + sizeof(gchar) * strlen(split_values[0]));
			g_stpcpy(ip_address, split_values[0]);
		}
		if(split_values[1] != NULL)
		{
			if(is_number(split_values[1]))
			{
				netmask = (gchar *) malloc (sizeof(gchar) * 16); 
				inttodd(netmask, cidrtoint(split_values[1]));
			}
		}
	}

	
	if (vlsm_counts)
	{
		if (ip_address && netmask)
		{
			if (verbose)
				print_info();
			vlsm_tree(ip_address, netmask, vlsm_counts);
		}
		else
			g_print("no ip address and/or netmask specified, use --help for usage information\n");
	}
	else if (do_host_count)
	{
		if (ip_address && netmask)
		{
			if (verbose)
				print_info();
			host_tree(ip_address, netmask, do_host_count);
		}
		else
			g_print("no ip address and/or netmask specified, use --help for usage information\n");
	}
	else if (do_net_count)
	{
		if (ip_address && netmask)
		{
			if (verbose)
				print_info();
			net_tree(ip_address, netmask, do_net_count);
		}
		else
			g_print("no ip address and/or netmask specified, use --help for usage information\n");
	}
	else if (do_summary)
	{
		if (verbose)
			print_info();
		net_summary();
	}
	else if (version)
	{
		if (verbose)
			print_info();
		else
			printf("%s\n", PACKAGE_STRING);
	}
	else
	{

		if (ip_address && netmask)
		{
			if (verbose)
				print_info();
			net_info(ip_address, netmask);
		}
		else
			g_print("%s", g_option_context_get_help(context, TRUE, NULL));
	}

	g_free(ip_address);
	g_free(netmask);
	return 0;
}

void print_info()
{
	fprintf(stderr, "%s Copyright (c) 2021 Matt Reath\n", PACKAGE_STRING);
	fprintf(stderr, "This program comes with ABSOLUTELY NO WARRANTY;\n");
	fprintf(stderr, "This is free software, and you are welcome to redistribute it\n");
	fprintf(stderr, "under certain circumstances. See the included COPYING file\n");
	fprintf(stderr, "for more information.\n\n");
}

void net_info(char *ip_address, char *mask)
{

	host h1;
	network n1;

	initialize_host(&h1, ip_address, mask);
	initialize_network(&n1, &h1);
	print_network_info(&n1, verbose);
}

void host_tree(char *ip_address, char *mask, int hosts)
{

	host h1;
	tnode *t1;

	initialize_host(&h1, ip_address, mask);

	t1 = (tnode *)malloc(sizeof(tnode));

	t1->right = NULL;
	t1->left = NULL;
	t1->parent = NULL;

	initialize_network(&t1->n, &h1);
	build_tree_host_count(t1, hosts);
	print_network_tree(t1, 0, verbose);
	free_network_tree(t1);
}

void net_tree(char *ip_address, char *mask, int nets)
{

	host h1;
	tnode *t1;

	initialize_host(&h1, ip_address, mask);

	t1 = (tnode *)malloc(sizeof(tnode));

	t1->right = NULL;
	t1->left = NULL;
	t1->parent = NULL;

	initialize_network(&t1->n, &h1);
	build_tree_net_count(t1, nets);
	print_network_tree(t1, 0, verbose);
	free_network_tree(t1);
}

void vlsm_tree(char *ip_address, char *mask, char *nets)
{

	host h1;
	tnode *t1;

	initialize_host(&h1, ip_address, mask);

	t1 = (tnode *)malloc(sizeof(tnode));

	t1->right = NULL;
	t1->left = NULL;
	t1->parent = NULL;

	initialize_network(&t1->n, &h1);

	//int j = 0;
	char *tok;
	char *sep = ",";

	for (tok = strtok(nets, sep); tok; tok = strtok(NULL, sep))
	{
		//verify host counts are legit
		if (!is_number(tok))
			g_error("vlsm string contains non-numeric values");

		// we should have left and right as an option here, set by CLI flag
		// default is left
		build_tree_vlsm(t1, atoi(tok), 0);
	}

	print_network_tree(t1, 0, verbose);
	free_network_tree(t1);
}

void net_summary()
{

	guint32 *networks = NULL;
	gchar* buffer;
	gchar* cidr_tok = "/";
	gchar* space_tok = " ";
	gchar* summarized_network;
	gchar* summarized_mask;
	guint32 val = 2147483648;
	size_t bufsize = 34;
	int i;
	int j;
	int k;
 
	// we can use just an array of unsigned 32-bit integers and
	// store the ip addresses in them, less memory usage and better
	// performance
	networks = (guint32*) g_malloc(sizeof(guint32) * MAX_NUM_TREES);

	/* 	read each line from stdin (using readline)
		check for "/" denoting a CIDR format
			if true, then split on "/" and initialize host
				need to develop an initialize host variant for CIDR
				need to develop cidrtoint utility function (convert CIDR to binary representation)
				check that mask is and integer and between 8 and 32 inside CIDRtoint
		if no "/" detected, check for space
			if true, split on space and initialize host as is done below */
			

	buffer = (gchar *) g_malloc(bufsize * sizeof(gchar));
	/* populate the array with the networks from the input */
	for (i = 0; getline(&buffer, &bufsize, stdin) > 0; i++)
	{
		if (buffer != NULL)
		{
			// remove newline characters from input
			int length = strlen(buffer);
			if (buffer[length-1] == '\n')
				buffer[length-1]  = '\0';

			if(g_strrstr(buffer,cidr_tok)) 
			{
				gchar** split_values = g_strsplit(buffer, cidr_tok, 2);
				networks[i] = ddtoint(split_values[0]);
				g_strfreev(split_values);
				
			}
			else if(g_strrstr(buffer,space_tok))
			{
				gchar** split_values = g_strsplit(buffer, space_tok, 2);
				networks[i] = ddtoint(split_values[0]);
				g_strfreev(split_values);
			}
		}
	}

	g_free(buffer);

	// *** NEW WAY O(n) much better performance ***
	// Sort by ip address -> create segments based on common bit patterns
	// -> summarize in one calculation all of the common bits in the segment
	// -> match mask to common bit boundry -> repeat for segments
	/* loop until there are no summarization left to do */

	guint32 common_bits = 0;
	guint32 mask_bits = 0;

	// loop through all networks
	for (j = 0; j < i; j++)
	{
		if (j == 0)
		{
			// this is our first iteration
			common_bits = networks[j];
		}
		else 
		{
			mask_bits = common_bits ^ networks[j];
			common_bits = common_bits & networks[j];
		}

	}


	summarized_network = (gchar *) g_malloc (sizeof(gchar) * 16); 
	inttodd(summarized_network, common_bits);

	for (k = 0 ; (mask_bits & val) != 2147483648; mask_bits = mask_bits << 1, k++ ) 
		;

	summarized_mask = (gchar *) g_malloc (sizeof(gchar) * 16); 
	inttodd(summarized_mask, get_mask_from_bits(k));
	
	if (verbose) {
		printf("%s\t%s\n", summarized_network, summarized_mask);
	}
	else 
	{
		printf("%s/%u\n", summarized_network, k);
	}
	

	// memory cleanup
	g_free(networks);
	g_free(summarized_network);
	g_free(summarized_mask);

	// *** OLD WAY O(n2) ***
	// while (made_changes)
	// {

	// 	made_changes = 0;
	// 	/* loop through each member */
	// 	for (j = 0; j < i; j++)
	// 	{
	// 		/* compare each member for summarization*/
	// 		for (k = 0; k < i; k++)
	// 		{

	// 			if ((j != k) && (t1 = combine_networks(networks[j], networks[k])) != NULL)
	// 			{
	// 				networks[j] = NULL;
	// 				networks[k] = NULL;

	// 				for (l = 0; networks[l] != NULL; l++)
	// 					;

	// 				networks[l] = t1;
	// 				made_changes = 1;
	// 			}
	// 		}
	// 	}
	// }

	// if(verbose)
	// 	printf("Networks were summarized as follows:\n\n");

	// for (i = 0; i < MAX_NUM_TREES; i++)
	// {
	// 	if (networks[i] != NULL)
	// 	{
	// 		if (verbose)
	// 		{
	// 			print_network_tree(networks[i], 0, TRUE);
	// 			printf("\n");
	// 		}
	// 		else
	// 		{
	// 			char ip_address[16];
	// 			inttodd(ip_address, networks[i]->n.address.ip_address);
	// 			printf("%s/%i\n", ip_address, get_bits_in_mask(networks[i]->n.address.mask));
	// 		}
	// 		free_network_tree(networks[i]);
	// 	}
	// }
}
